db/config.cc

```diff
@@ -990,7 +1010,6 @@ db::config::config(std::shared_ptr<db::extensions> exts)
             "In debug mode, report log-structured allocator sanitizer violations with a backtrace. Slow.")
     , flush_schema_tables_after_modification(this, "flush_schema_tables_after_modification", liveness::LiveUpdate, value_status::Used, true,
         "Flush tables in the system_schema keyspace after schema modification. This is required for crash recovery, but slows down tests and can be disabled for them")
-    , restrict_replication_simplestrategy(this, "restrict_replication_simplestrategy", liveness::LiveUpdate, value_status::Used, db::tri_mode_restriction_t::mode::FALSE, "Controls whether to disable SimpleStrategy replication. Can be true, false, or warn.")
```

Comment:
This never ends well. Even if we're 101% sure that this option is not used by anyone, once this rolls into prod clusters some of them may fail to (re)start.

And ~~if~~ when they will VladZ will be pissed off and PagerDuty will wake up someone at night. Not sure which is worse.

Reply:
The feature will generate a cql warning (visible through the cqlsh or a driver) - it's for on-premise OSS and enterprise + cloud. For cloud, devops should also configure the fail level, which will forbid to use SimpleStrategy at all (aligned with our SLA), but that we agreed to do only for the new cloud deployments, not when upgrading.  Guardrails + upgrades can lead to unforeseen errors, new deployments shouldn't (?).

Reply:
No no, I'm talking about completely different thing. If you start scylla with unknown option parsing throws and scylla just fails to start. Can you prove that there are no set-ups with `--restrict_replication_simplestrategy=SimpleStrategy` option hard-coded somewhere?

Reply:
Ah, OK. Good point. Of course I cannot prove it (because any customer could have configured it on his end), though AFAIK these restrictions have never been used/enabled and I don't see it mentioned anywhere in our scylladb github organization (I searched the whole codebase).
A couple of ideas what to do now:
1. just remove it, as we did with the old replication factor restriction - https://github.com/scylladb/scylladb/pull/14262/files#diff-9be2281a24993f93765810ad5298ac0e57ba9a58601b4025a09727f9fec331cdL1018
2. First obsolete, and only then remove. And also restore `minimum_keyspace_rf` (from the link above) and obsolete it as well.
3. Continue handling this configuration option, but just use my new implementation under the hood.
4. Keep both configuration parameters and let scylla's codebase rot and degrade ;)

Reply:
> If you start scylla with unknown option parsing throws and scylla just fails to start.

You sure, @xemul?

We had a case in the field where the `seeds` config option was misspelled as `seed` and not only scylla started up, it even picked itself (127.0.0.1) as the seed node and bootstrapped itself as a single node cluster (which then happily joined the rest of the cluster since it reused a dead node ip address instead of replacing it)

Reply:
Maybe there's a new GuardRail about unrecognized config options that I'm not familiar with.
If not, maybe there should be one... :)

Reply:
> You sure, @xemul?

I am
```
$ ./build/dev/scylla --unknown-option=true
Scylla version 5.5.0~dev-0.20231025.8e1ff745fae0 with build-id 4c465c4423bd12429dacfa788804d7b0171907d3 starting ...
command used: "./build/dev/scylla --unknown-option=true"
parsed command line options: [unknown-option: true]
error: unrecognised option '--unknown-option=true'
```

There's `value_status::Unused` thing for options that should not longer be used, but for which we still need to provide some handling (e.g. -- ignore)

Reply:
> AFAIK these restrictions have never been used/enabled and I don't see it mentioned anywhere in our scylladb github organization (I searched the whole codebase)

Well. Backward compatibility breakages are two-fold. If they come unnoticed you win, if they suddenly pop up the consequences can be disastrous. I wouldn't take that risk

Reply:
When I reverted `restrict_replication_simplestrategy`, I could run scylla with `--restrict-replication-simplestrategy false` and it started, but when I changed `value_status::Used` to `value_status::Unused`, I got the same error, as if this option was not available at all:
```
[piotrs@fedora scylla2]$ build/dev/scylla --restrict-replication-simplestrategy false
Scylla version 5.4.0~dev-0.20231001.93cec684f387 with build-id f2c6c3382a90ce95f04f9b0067c39a2ae253d3cd starting ...
command used: "build/dev/scylla --restrict-replication-simplestrategy false"
parsed command line options: [restrict-replication-simplestrategy, (positional) false]
error: unrecognised option '--restrict-replication-simplestrategy'

Try --help.
[piotrs@fedora scylla2]$ 
```
So this is not a way to go. Apparently marking it as `value_status::Invalid` works the same way as `value_status::Unused`. Am I doing it wrong?

Reply:
> @ptrsmrn handling of this option should be like 
> * mark it as Unused

Because of my uncertainty explained in a comment above, I left it as `value_status::Used` - please do correct me if I'm mistaken, and I'll change it to UNUSED.

> * on start -- check the option value is set and update the existing warn/throw lists with SimpleStrategy thus simulating the old behavior

What I did instead, was to revert the original logic, so modified `check_against_restricted_replication_strategies` and if this option is set, insert an entry either to the fail or warn lists.

> * on live-update -- also update the warn/throw list, warn in logs that the option is unused and going to be removed some day soon

Did only the first part, no warning is produced (yet)

> * ... wait some time/versions
> * remove the option completely

So because I'm not sure what to think about this `value_status::Unused`, I did only revert this option an am handling it to revert its logic. Other than that, I also added a cql-pytest to assert its really working. Will wait for feedback on how to continue.

Reply:
> When I reverted restrict_replication_simplestrategy, I could run scylla with --restrict-replication-simplestrategy false and it started ...

You mean -- when you _removed_ it and tried the scylla started, don't you? You must have just mixed up the binary, please double check

>  ... but when I changed value_status::Used to value_status::Unused, I got the same error, as if this option was not available at all ...

Confirmed and this must be a bug on its own. Here are some Unused options (there are much more than just that):

```
auto_adjust_flush_quota(this, "auto_adjust_flush_quota", value_status::Unused, false, ...
trickle_fsync(this, "trickle_fsync", value_status::Unused, false, ...
```

the former one is accepted, the latter generates CLI parsing error.

Reply:
BTW, based on the discussion around https://github.com/scylladb/scylladb/issues/15836 (also off GH), we want to reject any unrecognized option in scylla.yaml.

It seems like `Unused` options should be treated the same way, and we can have `Ignored` options as well for legacy options we may want to ignore, even if present in the config file.
And `Deprecated` like `replace_address` should be, with a message explaining what the user should use instead.

Reply:
> It seems like Unused options should be treated the same way, and we can have Ignored options as well for legacy options we may want to ignore, even if present in the config file.

Currently code comment says that Unused options are silently ignored. But I don't see the difference between Unused (legacy, parsed, but not shown in --help) vs Ignored(legacy, ignored and not shown in --help). They look the same, we just need a way to exclude an option from the `--help` output, but keep parsing it until we decide it's safe to just remove it.

Reply:
@xemul given your previous comment, do you think we could just leave `restrict_replication_simplestrategy` with `value_status::Used` and proceed?

Reply:
Well, I want to deprecate the `restrict_replication_simplestrategy` option in favor of new ones. For that the option should stop being visible in the `--help` output. For that, in turn, the `value_status::Unused` was invented, but as we see it's buggy and we (well -- me) don't understand what's wrong.

While I agree this PR and non-working Unused are independent issues, I'd still very much appreciate if someone spends time and sheds some light on why Unused bit doesn't work the way we expect.

Reply:
Would it be enough if I created a separate issue for the `value_status::Unused` and fix this guardrail once the value_status gets resolved?

Reply:
Please, mark the old option Unused in _this_ PR, we'll forget about it otherwise. And also please take a look at #15887 in the nearest time

Reply:
@xemul I'll just [repeat myself](https://github.com/scylladb/scylladb/pull/15399#discussion_r1373808796), if I set `restrict_replication_simplestrategy` to `value_status::Unused` and specify it via the cmd line, I get an error:
```
[piotrs@fedora scylla2]$ build/dev/scylla --restrict-replication-simplestrategy false
Scylla version 5.4.0~rc1-0.20231031.2aa29763af94 with build-id a27208086c3aba3b0fdc073ae742a3fae3aef05f starting ...
command used: "build/dev/scylla --blocked-reactor-notify-ms 10 --workdir workdir --smp 3 --developer-mode=yes --ring-delay-ms 3000 --skip-wait-for-gossip-to-settle 0 --memory 1G --restrict-replication-simplestrategy false"
parsed command line options: [blocked-reactor-notify-ms, (positional) 10, workdir: workdir, smp, (positional) 3, developer-mode: yes, ring-delay-ms: 3000, skip-wait-for-gossip-to-settle: 0, memory, (positional) 1G, restrict-replication-simplestrategy, (positional) false]
error: unrecognised option '--restrict-replication-simplestrategy'

Try --help.
[piotrs@fedora scylla2]$ ./tools/toolchain/dbuild ninja  -j25
```
Of course I tried it again, and is still not working for me. If I set it to `value_status::Used` and recompile, everything's fine. 
So that's the reason I left it as `Used`.

Reply:
@ptrsmrn , I'll also repeat (well, maybe rephrase) myself -- Unused options triggering parsing errors are BUGs and should be investigated. The --restrict-replication-simplestrategy option must be marked Unused. If we cannot do it without fixing the parsing BUG, then we first fix it, then merge this PR

Reply:
Got it. For now I changed the value to `value_status::Unused` and will investigate the bug.